### ISSUE TITLE
Understanding the output when returning probbilities

### ISSUE THREAD

**Issue by botasb** [datalogue/keras-attention#9]
Hi,
Thanks for this great job.
I don´t fully understand the shape of the return when returning probbilities. The shape is (self.timesteps, self.timesteps). In the visualization you squezze this to shape (0:output_length, 0:input_length), if I am not mistaken.
How do relatre these two? Could you elaborate a little bit?
I am trying to use it in a many-to-one rnn and it´s a little complicated to get useful visualization.
Thanks!

**Comment by zafarali** [comment]
Consider the following "attention vector", alpha, when trying to predict _something_.

| 5    |   | J   | A   | N   |      | 2    | 0 | 1    | 6   |
|------|---|-----|-----|-----|------|------|---|------|-----|
| 0.01 | 0 | 0.4 | 0.4 | 0.1 | 0.05 | 0.01 | 0 | 0.03 | 0.0 |

It's the result of Equation 2 (forget _t_ for now):
![image](https://user-images.githubusercontent.com/6295292/29527664-9b9f1f42-8667-11e7-8d0e-fa69f24b3eae.png)

Since the softmax usually characterizes a probability distribution over choices, we can interpret these are the _importance_ probabilities of the characters in making the output prediction. We see that "J", "A" and "N" are relatively more important than the others. 

Now we extend this idea to our application of translating our sequence. Here instead of predicting one thing, we need to make a sequence of predictions for each character in the output and we leverage these _importance_ scores to be able to weight the input sequence in our decision making.

That's equation 1, we are calculating these importance scores based on our previous prediction + the annotation. 

Technically speaking, the output shape should be (length of output sequence,  length of input sequence). However, since we are doing this using tensorflows static computation graphs, and considering that in this case the output sequence is at most the length of the input, I simplify to (length of input sequence, length of input sequence)

This is why in the visualization code I slice the array accordingly. 

From the minimal information I have, it seems like your many to one RNN will have a single attention vector as described above. If you provide more details, I may be able to help.

**Comment by botasb** [comment]
Got it. 
_Technically speaking, the output shape should be (length of output sequence, length of input sequence). However, since we are doing this using tensorflows static computation graphs, and considering that in this case the output sequence is at most the length of the input, I simplify to (length of input sequence, length of input sequence)_
That´s what I suspected. My mistake then. Visualization is good now (but not what I expected :)) 
Thanks a lot!

**Comment by zafarali** [comment]
Glad you're finding the code useful! Good luck
